scrambles/src/main/java/org/worldcubeassociation/tnoodle/puzzle/SquareOnePuzzle.java

```diff
@@ -196,32 +200,82 @@ protected int getRandomMoveCount() {
         return 40;
     }
 
-    /*
-    // TODO - we can't filter super aggresively until
-    // Chen Shuang's optimal solver is fixed.
     @Override
     protected String solveIn(PuzzleState ps, int n) {
-        FullCube f = ((SquareOneState)ps).toFullCube();
-        Search s = new Search();
-        String scramble = s.solutionOpt(f, n);
+        SquareOneState sqState = (SquareOneState) ps;
+
+        // apparently sq12phase can neither represent nor solve "unslashable" squares
+        if (!sqState.canSlash()) {
+            HashMap<String, SquareOneState> slashableSuccessors = sqState.getScrambleSuccessors();
+
+            Map.Entry<String, SquareOneState> bestSlashable = null;
+            int currentMin = Integer.MAX_VALUE;
+
+            for (Map.Entry<String, SquareOneState> possState : slashableSuccessors.entrySet()) {
+                Integer moveCost = topBottomCostsByMove.get(possState.getKey());
+
+                if (moveCost != null && moveCost < currentMin) {
+                    currentMin = moveCost;
+                    bestSlashable = possState;
+                }
+            }
+
+            // absolutely no successor found to make it slashable
+            if (bestSlashable == null) {
+                return null;
+            }
+
+            AlgorithmBuilder ab = new AlgorithmBuilder(this, AlgorithmBuilder.MergingMode.CANONICALIZE_MOVES, sqState);
+
+            try {
+                ab.appendMove(bestSlashable.getKey());
+
+                String nextBestSolution = bestSlashable.getValue().solveIn(n);
+                // initially just hope that it cancels with the n-move optimal solution
+                ab.appendAlgorithm(nextBestSolution);
+
+                if (ab.getTotalCost() > n) {
+                    // didn't cancel, try and find optimal solution in n-1
+                    ab = new AlgorithmBuilder(this, AlgorithmBuilder.MergingMode.CANONICALIZE_MOVES, sqState);
+                    ab.appendMove(bestSlashable.getKey());
+
+                    String oneMoveLessSolution = bestSlashable.getValue().solveIn(n - 1);
+                    // no need for any further checks here.
+                    // the "worst" thing that can happen is the n-1 move solution happens to cancel
+                    // (for whatever reason), in which case we've kept the bounds anyways
+                    ab.appendAlgorithm(oneMoveLessSolution);
+                }
+
+                return ab.getStateAndGenerator().generator;
+            } catch (InvalidMoveException e) {
+                e.printStackTrace();
+            }
+
+            // some invalid move exception happened; normally we don't ever get here
+            return null;
+        }
+
+        FullCube f = sqState.toFullCube();
+        String scramble = twoPhaseSearcher.get().solutionOpt(f, n);
         return scramble == null ? null : scramble.trim();
     }
-    */
 
     static HashMap<String, Integer> costsByMove = new HashMap<String, Integer>();
```

Comment:
If I understand this correctly, this holds the "old" mapping for costs that was not the one defined in the WCA Regulations.
I think we should remove the variable completely, as it's not meant to be used.

Reply:
There are two maps. One holds the "constant 1" count for any (x, y) as per [WCA Regulations](https://www.worldcubeassociation.org/regulations/#12c4) which is also the one being (correctly!) used for scramble cost.

The other map which holds cost "x+y" for (x,y) is here for a very specific reason: Since this change requires to make any SQ1-state "slashable" before passing it to the cs solver, we need to pick one such move to achieve "slashability".
Now take a solved SQ1 and apply move (-1, 0). We then have plenty of options to choose from, for example we might do (1, 0) or (4, 0) or (5, 0). In order to help the scrambler apply the "smallest" change possible, we keep this extra cost map (we could also parse the string but that would be *much* more resource-intensive)

Reply:
I totally got tricked!
Thanks for the detailed explanation!
Would you mind putting something similar in the comment, like right above the slashabilityCostsByMove, clarifying the purpose of this costs map?
There are relatively few comments in the code, so the more we detail how it works, the better it will be for someone to review/audit it later if it ever happens.